### ISSUE TITLE
LayerNorm casts input to float32 regardless of dtype

### ISSUE THREAD

**Issue by tomweingarten** [google/flax#612]
`    x = jnp.asarray(x, jnp.float32)`

Is this intentional? The LayerNorm module casts the input to float32 but casts other parameters to dtype, then multiplies them together. Naively it seems to me that this would be less efficient than casting the input to dtype.

**Comment by jheek** [comment]
Empirically we find that normalization layers (in particular computing moments) is somewhat sensitive to numerical precision issues. So we use float32 precision instead of float16/bfloat16 in half precision modules to avoid instability issues. Note that computing moments is not a big part of compute & memory compared to linear/conv modules so impact on performance is small.

**Comment by tomweingarten** [comment]
Thanks for that explanation!
